sqlmesh/core/engine_adapter/base.py

```diff
@@ -300,19 +301,62 @@ def replace_query(
                 Expected to be ordered to match the order of values in the dataframe.
             kwargs: Optional create table properties.
         """
-        table = exp.to_table(table_name)
+        target_table = exp.to_table(table_name)
         source_queries, columns_to_types = self._get_source_queries_and_columns_to_types(
-            query_or_df, columns_to_types, target_table=table_name
+            query_or_df, columns_to_types, target_table=target_table
         )
-        return self._create_table_from_source_queries(
-            table,
-            source_queries,
-            columns_to_types,
-            replace=True,
-            table_description=table_description,
-            column_descriptions=column_descriptions,
-            **kwargs,
+        columns_to_types = columns_to_types or self.columns(target_table)
+        query = source_queries[0].query_factory()
+        self_referencing = any(
+            quote_identifiers(table) == quote_identifiers(target_table)
+            for table in query.find_all(exp.Table)
         )
+        # If a query references itself then it must have a table created regardless of approach used.
+        if self_referencing:
+            self._create_table_from_columns(
```

Comment:
Yeah but that is fine right? Since it self-references then it would do a `CREATE OR REPLACE AS SELECT...` and reference itself so it would actually run. Without this it would error since it would reference a table that doesn't exist.

Reply:
what happened before this was added?

Reply:
Looks like it would error if self-referencing: https://github.com/TobikoData/sqlmesh/blob/c0bbe57041b090bd8c802e8ca04174040d8ba2f8/sqlmesh/core/engine_adapter/base.py#L618-L626

Reply:
Note that we wouldn't see this within SQLMesh since we always have a table but this now properly ensures that is the case regardless of context.